File: packages/llm-gateway/src/router/policy-engine.ts
import { readFileSync } from "fs";
import { resolve } from "path";
import { parse as parseYaml } from "yaml";
import type { SensitivityTier, ProviderName } from "../types.js";
import { ProviderError } from "@nexus/shared";

interface TierPolicy {
  primary?: ProviderName;
  secondary?: ProviderName;
  tertiary?: ProviderName;
  fallback?: ProviderName;
  blocked?: boolean;
}

interface PolicyFile {
  profile: string;
  routing: {
    tier_0: TierPolicy;
    tier_1: TierPolicy;
    tier_2: TierPolicy;
    tier_3: TierPolicy;
  };
  features?: {
    allow_cloud_tier2?: boolean;
    allow_cloud_tier3?: boolean;
    dlp_enabled?: boolean;
  };
}

export interface RoutingDecision {
  providers: ProviderName[];
  blocked: boolean;
  reason?: string;
}

/**
 * Interfaccia minimale per il client DB (lib `postgres`). Evita una dipendenza
 * diretta dal package gateway sul driver: il chiamante (nexus-gateway) inietta
 * il proprio `sql` gia' configurato. La firma e' volutamente lasca (return
 * `any`) per restare strutturalmente compatibile con i numerosi overload del
 * tipo `Sql` di `postgres` (tagged template + helper); il consumo concreto e'
 * tipizzato a destinazione in `refreshDbOverrides`.
 */
// Return volutamente lasco (`any`): il client `postgres` espone numerosi
// overload (tagged template + helper `Sql`) che non si lasciano vincolare a un
// `Promise<T[]>` esatto. Il tipo concreto delle righe e' applicato in
// `refreshDbOverrides` tramite cast. Il package non ha lint eslint attivo
// (nessuno script `lint` in package.json), quindi `any` qui e' accettabile e
// confinato a questo confine di iniezione DB.
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type SettingsDb = (strings: TemplateStringsArray, ...values: unknown[]) => any;

const PARSE_BOOL = (v: string | undefined): boolean | undefined => {
  if (v === undefined || v === null) return undefined;
  const t = String(v).trim().toLowerCase();
  if (t === "true" || t === "1") return true;
  if (t === "false" || t === "0") return false;
  return undefined;
};

export class PolicyEngine {
  private policy: PolicyFile;

  // Provider considerati cloud (inviano dati a servizi esterni). Fonte unica
  // usata sia per il filtro tenant sia per l'enforcement dei flag DLP per-tier.
  private static readonly CLOUD_PROVIDERS: ProviderName[] = [
    "anthropic",
    "openai",
    "mistral",
    "deepseek",
    "google",
  ];

  // Override letti dai settings DB (regola G: configurazione dal DB, non dal
  // file YAML). `undefined` = nessun valore DB letto -> usa il default YAML.
  private dbOverrides: {
    allow_cloud_tier2?: boolean;
    allow_cloud_tier3?: boolean;
    dlp_enabled?: boolean;
  } = {};
  private dbOverridesFetchedAt = 0;
  private static readonly DB_TTL_MS = 60_000;

  constructor(policyFile: string) {
    const raw = readFileSync(resolve(policyFile), "utf-8");
    this.policy = parseYaml(raw) as PolicyFile;
  }

  /**
   * Valore effettivo di un flag: priorita' al DB (settings), fallback YAML.
   */
  private flag(key: "allow_cloud_tier2" | "allow_cloud_tier3" | "dlp_enabled"): boolean | undefined {
    const fromDb = this.dbOverrides[key];
    if (fromDb !== undefined) return fromDb;
    return this.policy.features?.[key];
  }

  /**
   * Indica se il tier richiede l'enforcement del flag cloud DLP e qual e' il
   * flag corrispondente. Tier 0/1 non hanno gate cloud (sempre permessi).
   */
  private cloudGateForTier(tier: SensitivityTier): boolean | undefined {
    if (tier >= 3) return this.flag("allow_cloud_tier3");
    if (tier === 2) return this.flag("allow_cloud_tier2");
    return undefined; // tier 0/1: nessun gate cloud
  }

  /**
   * Ricarica i flag DLP dai settings DB con cache TTL 60s. Se il DB e'
   * irraggiungibile mantiene i valori correnti (fallback graceful sullo YAML).
   * `force` ignora la cache (usato all'avvio).
   */
  async refreshDbOverrides(db: SettingsDb | null | undefined, force = false): Promise<void> {
    if (!db) return;
    const now = Date.now();
    if (!force && now - this.dbOverridesFetchedAt < PolicyEngine.DB_TTL_MS) return;

    try {
      const rows = (await db`
        SELECT key, value FROM settings
        WHERE key IN ('dlp_allow_cloud_tier2', 'dlp_allow_cloud_tier3', 'dlp_enabled')
      `) as Array<{ key: string; value: string }>;
      const map = new Map(rows.map((r) => [r.key, r.value]));
      const next: typeof this.dbOverrides = {};
      next.allow_cloud_tier2 = PARSE_BOOL(map.get("dlp_allow_cloud_tier2"));
      next.allow_cloud_tier3 = PARSE_BOOL(map.get("dlp_allow_cloud_tier3"));
      next.dlp_enabled = PARSE_BOOL(map.get("dlp_enabled"));
      this.dbOverrides = next;
      this.dbOverridesFetchedAt = now;
    } catch (err) {
      // DB down: non azzerare gli override gia' noti; il routing prosegue sui
      // valori YAML / ultimi noti. Timestamp non aggiornato -> retry al prossimo giro.
      console.warn(
        `[policy-engine] refreshDbOverrides fallito (${(err as Error).message}). ` +
          `Mantengo i flag DLP correnti (fallback YAML/ultimi noti).`
      );
    }
  }

  /**
   * True se la classificazione DLP e' disattivata da settings (`dlp_enabled=false`).
   * In tal caso il caller deve trattare ogni richiesta come tier 0.
   */
  isDlpDisabled(): boolean {
    return this.flag("dlp_enabled") === false;
  }

  decide(
    tier: SensitivityTier,
    feature: string,
    tenantFlags: Record<string, boolean> = {}
  ): RoutingDecision {
    // DLP disattivato da DB: nessuna sensibilita', si instrada sempre come tier 0.
    const effectiveTier = this.isDlpDisabled() ? (0 as SensitivityTier) : tier;

    const tierKey = `tier_${effectiveTier}` as keyof PolicyFile["routing"];
    const tierPolicy = this.policy.routing?.[tierKey];

    if (!tierPolicy) {
      return {
        providers: [],
        blocked: true,
        reason: `Nessuna policy per tier ${effectiveTier}`,
      };
    }

    // Gate cloud per-tier: la fonte di verita' del blocco cloud e' il flag DLP
    // (settings DB `dlp_allow_cloud_tierN`, fallback YAML `features.*`), NON il
    // campo struttura `tier_N.blocked`. Se il flag DLP e' definito, vince:
    //   - flag = false  -> blocca i provider cloud per questo tier
    //   - flag = true   -> consente i provider cloud, ignorando `tier_N.blocked`
    // Se il flag non e' definito (ne' DB ne' YAML), si ricade sul comportamento
    // storico basato su `tier_N.blocked`.
    const cloudGate = this.cloudGateForTier(effectiveTier);
    const blockCloudByDlp = cloudGate === false;

    if (cloudGate === undefined && tierPolicy.blocked) {
      return {
        providers: [],
        blocked: true,
        reason: `Tier ${effectiveTier} bloccato dalla policy (profilo: ${this.policy.profile})`,
      };
    }

    // Tenant override: se il tenant ha allow_cloud disabilitato, esclude provider cloud
    const tenantBlocksCloud = tenantFlags["block_cloud"] === true;
    const cloudProviders = PolicyEngine.CLOUD_PROVIDERS;

    const ordered: ProviderName[] = [
      tierPolicy.primary,
      tierPolicy.secondary,
      tierPolicy.tertiary,
      tierPolicy.fallback,
    ].filter((n): n is ProviderName => !!n);

    const excludeCloud = tenantBlocksCloud || blockCloudByDlp;
    const filtered = excludeCloud
      ? ordered.filter((p) => !cloudProviders.includes(p))
      : ordered;

    if (filtered.length === 0) {
      return {
        providers: [],
        blocked: true,
        reason: blockCloudByDlp
          ? `Tier ${effectiveTier}: provider cloud bloccati dal flag DLP (dlp_allow_cloud_tier${effectiveTier}=false) e nessun provider locale configurato`
          : `Nessun provider disponibile per tier ${effectiveTier} con i flag tenant correnti`,
      };
    }

    return { providers: filtered, blocked: false };
  }

  getProfile(): string {
    return this.policy.profile;
  }

  // Confronta il tier dichiarato col tier rilevato dal classifier.
  //
  // NON blocca piu' con 400 quando detected > claimed: il caller (mcp-core /
  // brain) NON puo' conoscere a priori il tier di sensibilita' del contenuto
  // — e' proprio il gateway a rilevarlo via classifier. Bloccare il caller per
  // non aver "indovinato" il tier rompeva ogni richiesta con PII (es. specifiche
  // applicative che menzionano email/anagrafica), perche' il default dichiarato
  // e' 0. La gestione corretta e' l'auto-elevazione: `complete`/`preview`
  // calcolano gia' `effectiveTier = max(claimed, detected)` e instradano ai
  // provider permessi per quel tier (le restrizioni di policy — es. cloud
  // tier3 — restano applicate da `decideTier`/`buildFallbackChain`). Qui ci
  // limitiamo a segnalare l'elevazione per audit, senza interrompere il flusso.
  validateTierClaim(claimed: SensitivityTier, detected: SensitivityTier): void {
    if (detected > claimed) {
      console.warn(
        `[policy-engine] tier auto-elevato: dichiarato ${claimed} -> rilevato ${detected}. ` +
          `Routing applicato sul tier effettivo ${detected} (le policy del tier restano enforced).`
      );
    }
  }
}

File: crates/nexus-gateway/src/policy_engine.rs
//! Motore di policy di routing per tier di sensibilita'.
//!
//! Porting fedele di `packages/llm-gateway/src/router/policy-engine.ts`. Carica
//! la policy YAML (`config/policies/<profilo>.yaml`): profilo, routing
//! tier_0..tier_3 (primary/secondary/tertiary/fallback/blocked) e i flag DLP
//! (`features.allow_cloud_tier2/3`, `dlp_enabled`).
//!
//! Decisione di routing (`decide`):
//!   - se `dlp_enabled=false`, ogni richiesta e' trattata come tier 0;
//!   - gate cloud per-tier governato dal flag DLP (DB > YAML), non da `blocked`;
//!   - tenant override `block_cloud` e gate DLP escludono i provider cloud;
//!   - ritorna la lista ordinata dei provider per il tier effettivo.
//!
//! Override DB (regola G): i flag `dlp_allow_cloud_tier2/3` e `dlp_enabled`
//! arrivano dai `settings`, con priorita' sul valore YAML. La cache TTL e' il
//! punto unico `nexus_cache::TtlCache` (regola L): niente timer/contatori
//! manuali. Se il DB e' irraggiungibile gli override correnti restano validi
//! (fallback graceful sullo YAML / ultimi valori noti).

use std::sync::Arc;
use std::time::Duration;

use nexus_cache::TtlCache;
use serde::Deserialize;
use sqlx::PgPool;

use crate::types::SensitivityTier;

/// Policy di un singolo tier nel file YAML (`tier_N`).
#[derive(Debug, Clone, Default, Deserialize)]
struct TierPolicy {
    #[serde(default)]
    primary: Option<String>,
    #[serde(default)]
    secondary: Option<String>,
    #[serde(default)]
    tertiary: Option<String>,
    #[serde(default)]
    fallback: Option<String>,
    #[serde(default)]
    blocked: bool,
}

/// Blocco `routing` del file policy.
#[derive(Debug, Clone, Default, Deserialize)]
struct RoutingTable {
    #[serde(default)]
    tier_0: TierPolicy,
    #[serde(default)]
    tier_1: TierPolicy,
    #[serde(default)]
    tier_2: TierPolicy,
    #[serde(default)]
    tier_3: TierPolicy,
}

impl RoutingTable {
    /// Restituisce la policy del tier indicato (0..3). `None` per tier fuori range.
    fn for_tier(&self, tier: SensitivityTier) -> Option<&TierPolicy> {
        match tier {
            0 => Some(&self.tier_0),
            1 => Some(&self.tier_1),
            2 => Some(&self.tier_2),
            3 => Some(&self.tier_3),
            _ => None,
        }
    }
}

/// Blocco `features` del file policy (default DLP).
#[derive(Debug, Clone, Default, Deserialize)]
struct PolicyFeatures {
    #[serde(default)]
    allow_cloud_tier2: Option<bool>,
    #[serde(default)]
    allow_cloud_tier3: Option<bool>,
    #[serde(default)]
    dlp_enabled: Option<bool>,
}

/// File policy completo. Sono dichiarati solo i campi usati dal routing; gli
/// altri blocchi (providers, redaction, telemetry, ...) sono ignorati.
#[derive(Debug, Clone, Default, Deserialize)]
struct PolicyFile {
    #[serde(default)]
    profile: String,
    #[serde(default)]
    routing: RoutingTable,
    #[serde(default)]
    features: PolicyFeatures,
}

/// Esito di una decisione di routing (`RoutingDecision` del TS).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RoutingDecision {
    /// Provider ordinati per priorita' (primary -> fallback) ammessi.
    pub providers: Vec<String>,
    /// `true` se nessun provider e' instradabile per la richiesta.
    pub blocked: bool,
    /// Motivo del blocco / dell'esito (per audit).
    pub reason: Option<String>,
    /// SEGNALE STRUTTURATO (regola M): `true` quando il blocco deriva dal gate
    /// DLP cloud per-tier (contenuto riservato -> provider cloud esclusi), NON
    /// da tier ignoto o flag tenant. I caller lo usano per rispondere col
    /// codice dedicato `POLICY_TIER_EXCLUDED` invece del generico
    /// `TIER_BLOCKED`, senza parsare la `reason` testuale.
    pub dlp_blocked: bool,
}

/// Override DLP letti dai settings DB. `None` su un campo = nessun valore in DB
/// -> si ricade sul valore YAML.
#[derive(Debug, Clone, Copy, Default)]
struct DlpOverrides {
    allow_cloud_tier2: Option<bool>,
    allow_cloud_tier3: Option<bool>,
    dlp_enabled: Option<bool>,
}

/// Errore di caricamento policy.
#[derive(Debug, thiserror::Error)]
pub enum PolicyError {
    #[error("caricamento policy fallito: {0}")]
    Load(String),
}

/// Provider considerati cloud (inviano dati a servizi esterni). Fonte unica per
/// il filtro tenant e per l'enforcement dei flag DLP per-tier. Non sono "nomi di
/// modello" (regola G), ma l'insieme strutturale dei provider non-locali, allineato
/// alla costante `CLOUD_PROVIDERS` del TS.
const CLOUD_PROVIDERS: [&str; 5] = ["anthropic", "openai", "mistral", "deepseek", "google"];

/// TTL della cache override DB (60s, come il TS `DB_TTL_MS`).
const DB_TTL: Duration = Duration::from_secs(60);

/// Motore di policy. La policy YAML e' immutabile dopo il caricamento; gli
/// override DB vivono in una `TtlCache` condivisa (clonabile a basso costo).
#[derive(Debug, Clone)]
pub struct PolicyEngine {
    policy: Arc<PolicyFile>,
    // Chiave unit: c'e' un solo set di override globale, ma la scadenza temporale
    // e' gestita dal punto unico TtlCache (regola L) invece che da un timestamp
    // manuale. `get(&())` ritorna gli override solo se non scaduti.
    overrides: TtlCache<(), DlpOverrides>,
}

impl PolicyEngine {
    /// Carica la policy dal file YAML indicato.
    pub fn from_yaml_file(path: &str) -> Result<Self, PolicyError> {
        let raw = std::fs::read_to_string(path).map_err(|e| PolicyError::Load(e.to_string()))?;
        Self::from_yaml_str(&raw)
    }

    /// Carica la policy da una stringa YAML (usato nei test e dal loader file).
    pub fn from_yaml_str(yaml: &str) -> Result<Self, PolicyError> {
        let policy: PolicyFile =
            serde_yaml::from_str(yaml).map_err(|e| PolicyError::Load(e.to_string()))?;
        Ok(Self {
            policy: Arc::new(policy),
            overrides: TtlCache::new(DB_TTL),
        })
    }

    /// Nome del profilo policy (`getProfile` del TS).
    pub fn profile(&self) -> &str {
        &self.policy.profile
    }

    /// Override correnti (se in cache e non scaduti), altrimenti default vuoto.
    fn current_overrides(&self) -> DlpOverrides {
        self.overrides.get(&()).unwrap_or_default()
    }

    /// Valore effettivo di un flag: priorita' al DB (settings), fallback YAML.
    fn flag(&self, key: DlpFlag) -> Option<bool> {
        let ov = self.current_overrides();
        let (from_db, from_yaml) = match key {
            DlpFlag::AllowCloudTier2 => (ov.allow_cloud_tier2, self.policy.features.allow_cloud_tier2),
            DlpFlag::AllowCloudTier3 => (ov.allow_cloud_tier3, self.policy.features.allow_cloud_tier3),
            DlpFlag::DlpEnabled => (ov.dlp_enabled, self.policy.features.dlp_enabled),
        };
        from_db.or(from_yaml)
    }

    /// Gate cloud per-tier. Tier 0/1 non hanno gate (sempre permessi -> `None`).
    fn cloud_gate_for_tier(&self, tier: SensitivityTier) -> Option<bool> {
        if tier >= 3 {
            self.flag(DlpFlag::AllowCloudTier3)
        } else if tier == 2 {
            self.flag(DlpFlag::AllowCloudTier2)
        } else {
            None
        }
    }

    /// `true` se la classificazione DLP e' disattivata (`dlp_enabled=false`).
    /// In tal caso il caller tratta ogni richiesta come tier 0.
    pub fn is_dlp_disabled(&self) -> bool {
        self.flag(DlpFlag::DlpEnabled) == Some(false)
    }

    /// `true` se i provider CLOUD sono bloccati dal gate DLP per il tier dato
    /// (flag `dlp_allow_cloud_tier2/3`, DB > YAML). Tier 0/1 non hanno gate.
    /// Punto unico (regola L) riusato dal path pin di `routes.rs`: il pin
    /// bypassa il ROUTING (`decide`) ma non questo gate di sicurezza.
    pub fn cloud_blocked_for_tier(&self, tier: SensitivityTier) -> bool {
        if self.is_dlp_disabled() {
            return false;
        }
        self.cloud_gate_for_tier(tier) == Some(false)
    }

    /// `true` se `name` e' un provider cloud (invia dati a servizi esterni).
    /// Espone la fonte unica [`CLOUD_PROVIDERS`] ai caller del gate.
    pub fn is_cloud_provider(name: &str) -> bool {
        CLOUD_PROVIDERS.contains(&name)
    }

    /// Ricarica i flag DLP dai settings DB con cache TTL 60s. Se il DB e'
    /// irraggiungibile mantiene i valori correnti (fallback graceful sullo YAML).
    /// `force=true` ignora la cache (usato all'avvio).
    ///
    /// La gestione TTL e' delegata interamente a `TtlCache`: se la entry e'
    /// ancora valida e non si forza, si esce subito senza toccare il DB.
    pub async fn refresh_db_overrides(&self, pool: &PgPool, force: bool) {
        if !force && self.overrides.get(&()).is_some() {
            return;
        }

        // settings.value e' TEXT; leggiamo le 3 chiavi DLP in un colpo.
        let rows = sqlx::query_as::<_, (String, String)>(
            "SELECT key, value FROM settings \
             WHERE key IN ('dlp_allow_cloud_tier2', 'dlp_allow_cloud_tier3', 'dlp_enabled')",
        )
        .fetch_all(pool)
        .await;

        match rows {
            Ok(rows) => {
                let mut next = DlpOverrides::default();
                for (key, value) in &rows {
                    let parsed = parse_bool(value);
                    match key.as_str() {
                        "dlp_allow_cloud_tier2" => next.allow_cloud_tier2 = parsed,
                        "dlp_allow_cloud_tier3" => next.allow_cloud_tier3 = parsed,
                        "dlp_enabled" => next.dlp_enabled = parsed,
                        _ => {}
                    }
                }
                self.overrides.insert((), next);
            }
            Err(err) => {
                // DB down: non azzerare gli override gia' noti. La entry scaduta
                // resta tale -> retry al prossimo giro; il routing prosegue sui
                // valori YAML / ultimi noti.
                tracing::warn!(
                    error = %err,
                    "policy-engine: refresh override DLP fallito, mantengo i flag correnti (fallback YAML)"
                );
            }
        }
    }

    /// Decide la lista di provider per (tier, feature, tenant flags).
    /// Porting 1:1 della `decide` del TS.
    pub fn decide(
        &self,
        tier: SensitivityTier,
        _feature: &str,
        tenant_flags: &std::collections::HashMap<String, bool>,
    ) -> RoutingDecision {
        // DLP disattivato da DB: nessuna sensibilita', si instrada come tier 0.
        let effective_tier: SensitivityTier = if self.is_dlp_disabled() { 0 } else { tier };

        let Some(tier_policy) = self.policy.routing.for_tier(effective_tier) else {
            return RoutingDecision {
                providers: Vec::new(),
                blocked: true,
                reason: Some(format!("Nessuna policy per tier {effective_tier}")),
                dlp_blocked: false,
            };
        };

        // Gate cloud per-tier: il flag DLP (DB > YAML) e' la fonte di verita' del
        // blocco cloud, NON il campo `tier_N.blocked`. Se il flag e' definito vince;
        // se non definito (ne' DB ne' YAML) si ricade sul comportamento storico
        // basato su `blocked`.
        let cloud_gate = self.cloud_gate_for_tier(effective_tier);
        let block_cloud_by_dlp = cloud_gate == Some(false);

        if cloud_gate.is_none() && tier_policy.blocked {
            return RoutingDecision {
                providers: Vec::new(),
                blocked: true,
                reason: Some(format!(
                    "Tier {effective_tier} bloccato dalla policy (profilo: {})",
                    self.policy.profile
                )),
                dlp_blocked: false,
            };
        }

        // Tenant override: se il tenant blocca il cloud, esclude i provider cloud.
        let tenant_blocks_cloud = tenant_flags.get("block_cloud").copied().unwrap_or(false);

        let ordered: Vec<String> = [
            tier_policy.primary.as_deref(),
            tier_policy.secondary.as_deref(),
            tier_policy.tertiary.as_deref(),
            tier_policy.fallback.as_deref(),
        ]
        .into_iter()
        .flatten()
        .map(|s| s.to_string())
        .collect();

        let exclude_cloud = tenant_blocks_cloud || block_cloud_by_dlp;
        let filtered: Vec<String> = if exclude_cloud {
            ordered
                .into_iter()
                .filter(|p| !CLOUD_PROVIDERS.contains(&p.as_str()))
                .collect()
        } else {
            ordered
        };

        if filtered.is_empty() {
            let reason = if block_cloud_by_dlp {
                format!(
                    "Tier {effective_tier}: provider cloud bloccati dal flag DLP \
                     (dlp_allow_cloud_tier{effective_tier}=false) e nessun provider locale configurato"
                )
            } else {
                format!("Nessun provider disponibile per tier {effective_tier} con i flag tenant correnti")
            };
            return RoutingDecision {
                providers: Vec::new(),
                blocked: true,
                reason: Some(reason),
                dlp_blocked: block_cloud_by_dlp,
            };
        }

        RoutingDecision {
            providers: filtered,
            blocked: false,
            reason: None,
            dlp_blocked: false,
        }
    }

    /// Segnala (solo per audit) un'auto-elevazione di tier. Non interrompe il
    /// flusso: e' il gateway a rilevare il tier reale via classifier, quindi un
    /// `detected > claimed` non e' un errore del caller (vedi nota nel TS).
    pub fn validate_tier_claim(&self, claimed: SensitivityTier, detected: SensitivityTier) {
        if detected > claimed {
            tracing::warn!(
                claimed,
                detected,
                "policy-engine: tier auto-elevato; routing applicato sul tier effettivo (policy enforced)"
            );
        }
    }
}

/// Flag DLP gestiti dal motore.
#[derive(Debug, Clone, Copy)]
enum DlpFlag {
    AllowCloudTier2,
    AllowCloudTier3,
    DlpEnabled,
}

/// Parsing booleano tollerante (`PARSE_BOOL` del TS): "true"/"1" -> true,
/// "false"/"0" -> false, altro -> `None` (valore ignoto -> nessun override).
fn parse_bool(v: &str) -> Option<bool> {
    match v.trim().to_ascii_lowercase().as_str() {
        "true" | "1" => Some(true),
        "false" | "0" => Some(false),
        _ => None,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashMap;

    // Policy minimale che riproduce la struttura di default.yaml ma senza i
    // blocchi non usati dal routing. Autosufficiente: nessuna lettura on-disk.
    const POLICY: &str = r#"
profile: cloud
features:
  allow_cloud_tier2: true
  allow_cloud_tier3: false
  dlp_enabled: true
routing:
  tier_0:
    primary: openai
    secondary: deepseek
    tertiary: google
    fallback: mistral
  tier_1:
    primary: openai
    secondary: deepseek
  tier_2:
    primary: openai
    secondary: deepseek
  tier_3:
    primary: anthropic
    secondary: openai
    fallback: deepseek
"#;

    fn engine() -> PolicyEngine {
        PolicyEngine::from_yaml_str(POLICY).expect("policy valida")
    }

    #[test]
    fn tier_0_ritorna_chain_ordinata() {
        let e = engine();
        let d = e.decide(0, "chat", &HashMap::new());
        assert!(!d.blocked);
        assert_eq!(d.providers, vec!["openai", "deepseek", "google", "mistral"]);
    }

    #[test]
    fn tier_3_cloud_bloccato_da_flag_yaml() {
        let e = engine();
        // allow_cloud_tier3=false (YAML) -> tutti i provider del tier 3 sono cloud
        // -> nessun provider locale -> blocked.
        let d = e.decide(3, "chat", &HashMap::new());
        assert!(d.blocked);
        assert!(d.providers.is_empty());
        // Segnale strutturato: il blocco viene dal gate DLP (regola M).
        assert!(d.dlp_blocked);
        assert!(d.reason.unwrap().contains("dlp_allow_cloud_tier3=false"));
        // Gate riusato dal path pin: cloud bloccato a tier 3, libero a tier 0-2.
        assert!(e.cloud_blocked_for_tier(3));
        assert!(!e.cloud_blocked_for_tier(2));
        assert!(!e.cloud_blocked_for_tier(0));
        assert!(PolicyEngine::is_cloud_provider("deepseek"));
        assert!(!PolicyEngine::is_cloud_provider("vllm"));
    }

    #[test]
    fn tier_2_cloud_permesso_da_flag_yaml() {
        let e = engine();
        // allow_cloud_tier2=true -> i provider cloud restano.
        let d = e.decide(2, "chat", &HashMap::new());
        assert!(!d.blocked);
        assert_eq!(d.providers, vec!["openai", "deepseek"]);
    }

    #[test]
    fn tenant_block_cloud_esclude_provider_cloud() {
        let e = engine();
        let mut flags = HashMap::new();
        flags.insert("block_cloud".to_string(), true);
        // Tier 0 ha solo provider cloud nello YAML di test -> tutti esclusi -> blocked.
        let d = e.decide(0, "chat", &flags);
        assert!(d.blocked);
        assert!(d.providers.is_empty());
        // Blocco da flag TENANT, non dal gate DLP: dlp_blocked resta false.
        assert!(!d.dlp_blocked);
    }

    #[test]
    fn override_db_dlp_disabilitato_instrada_come_tier0() {
        let e = engine();
        // Simula override DB: dlp_enabled=false -> ogni tier trattato come tier 0.
        e.overrides.insert(
            (),
            DlpOverrides {
                dlp_enabled: Some(false),
                ..Default::default()
            },
        );
        assert!(e.is_dlp_disabled());
        // Richiesta tier 3, ma instradata come tier 0 (chain completa, non bloccata).
        let d = e.decide(3, "chat", &HashMap::new());
        assert!(!d.blocked);
        assert_eq!(d.providers, vec!["openai", "deepseek", "google", "mistral"]);
    }

    #[test]
    fn override_db_vince_su_yaml() {
        let e = engine();
        // YAML: allow_cloud_tier3=false. Override DB: true -> cloud tier 3 permesso.
        e.overrides.insert(
            (),
            DlpOverrides {
                allow_cloud_tier3: Some(true),
                ..Default::default()
            },
        );
        let d = e.decide(3, "chat", &HashMap::new());
        assert!(!d.blocked);
        assert_eq!(d.providers, vec!["anthropic", "openai", "deepseek"]);
    }

    #[test]
    fn parse_bool_tollerante() {
        assert_eq!(parse_bool("true"), Some(true));
        assert_eq!(parse_bool(" 1 "), Some(true));
        assert_eq!(parse_bool("FALSE"), Some(false));
        assert_eq!(parse_bool("0"), Some(false));
        assert_eq!(parse_bool("boh"), None);
    }
}
